lib/coffee_place/location.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+require_relative './result'
+require_relative './geo'
+require_relative './earth_distance'
+
+module CoffeePlace
+  # Represents a location on the map
+  class Location
+    include Geo
+    include EarthDistance
+
+    attr_accessor :lat, :lon, :name
+
+    def initialize(lat:, lon:, name:)
+      @lat = lat
+      @lon = lon
+      @name = name
+    end
+
+    def distance_to(other_location)
+      earth_distance(self, other_location)
+    end
+
+    def valid_latitude?
+      within_latitude_bounds?(@lat)
+    end
+
+    def valid_longitude?
+      within_longitude_bounds?(@lon)
+    end
+
+    def to_s
+      "location(lat = #{@lat}, lon: #{@lon}, name: #{@name})"
+    end
+
+    def ==(other)
+      same_place?(other) && (@name == other.name)
+    end
+
+    def same_place?(other)
+      (@lat == other.lat) && (@lon == other.lon)
+    end
+
+    class << self
+      def validate(lat:, lon:, name:)
+        result_lat = parse_coordinate(lat, 'latitude')
+        return result_lat unless result_lat.success?
+
+        result_lon = parse_coordinate(lon, 'longitude')
+        return result_lon unless result_lon.success?
+
+        location = new(lat: result_lat.value, lon: result_lon.value, name: name.to_s)
+
+        return Result.failure("Invalid latitude: #{lat}") unless location.valid_latitude?
+        return Result.failure("Invalid longitude: #{lon}") unless location.valid_longitude?
```

Comment:
What is your take on multiple return statements? Is it good / bad, why, when should one use it.

Reply:
^I've the same question. :smile:

Reply:
I like using guard clauses - it is the only place in Ruby where I use the `return` statment.

Basically the success path travels down the method, while guard clauses will end control flow prematurely
in case of an unexpected input, or some precondition not met.

Here, if the location is invalid, which should be unlikely, I fail-fast on the first error and return it.
Another approach would be to have an `@errors` array, in which to collect errors. 
I would then only check the errors at the end of the method and use Ruby's implicit expression return value
to return either the errors -if there are any, or the success value.

I think both approaches are valid.

I like Avdi Grimm's [Confident Ruby](https://pragprog.com/titles/agcr/confident-ruby/) book quite alot.
He uses guard clauses in order to clean up methods with branching logic in pretty neat ways.

Reply:
Update: fixed by 482136f9